server/src/main/java/org/opensearch/extensions/rest/RestActionsRequestHandler.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.extensions.rest;
+
+import org.opensearch.extensions.DiscoveryExtension;
+import org.opensearch.rest.RestController;
+import org.opensearch.rest.RestHandler;
+import org.opensearch.transport.TransportResponse;
+import org.opensearch.transport.TransportService;
+
+import java.util.Map;
+
+/**
+ * Handles requests to register extension REST actions.
+ *
+ * @opensearch.internal
+ */
+public class RestActionsRequestHandler {
+
+    private final RestController restController;
+    private final Map<String, DiscoveryExtension> extensionIdMap;
+    private final TransportService transportService;
+
+    /**
+     * Instantiates a new REST Actions Request Handler using the Node's RestController.
+     *
+     * @param restController  The Node's {@link RestController}.
+     * @param extensionIdMap  A map of extension uniqueId to DiscoveryExtension
+     * @param transportService  The Node's transportService
+     */
+    public RestActionsRequestHandler(
+        RestController restController,
+        Map<String, DiscoveryExtension> extensionIdMap,
+        TransportService transportService
+    ) {
+        this.restController = restController;
+        this.extensionIdMap = extensionIdMap;
+        this.transportService = transportService;
+    }
+
+    /**
+     * Handles a {@link RegisterRestActionsRequest}.
+     *
+     * @param restActionsRequest  The request to handle.
+     * @return A {@link RegisterRestActionsResponse} indicating success.
+     * @throws Exception if the request is not handled properly.
+     */
+    public TransportResponse handleRegisterRestActionsRequest(RegisterRestActionsRequest restActionsRequest) throws Exception {
+        DiscoveryExtension discoveryExtension = extensionIdMap.get(restActionsRequest.getUniqueId());
+        RestHandler handler = new RestSendToExtensionAction(restActionsRequest, discoveryExtension, transportService);
+        restController.registerHandler(handler);
```

Comment:
What if `registerHandler` fails? Do we handle it gracefully and the extension know about it?

Reply:
Yeah I did look at the method, looks like it doesnt throw an exception nor an error. 
That makes sense, thanks!

```
public void registerHandler(final RestHandler restHandler) {
        restHandler.routes().forEach(route -> registerHandler(route.getMethod(), route.getPath(), restHandler));
        restHandler.deprecatedRoutes()
            .forEach(route -> registerAsDeprecatedHandler(route.getMethod(), route.getPath(), restHandler, route.getDeprecationMessage()));
        restHandler.replacedRoutes()
            .forEach(
                route -> registerWithDeprecatedHandler(
                    route.getMethod(),
                    route.getPath(),
                    restHandler,
                    route.getDeprecatedMethod(),
                    route.getDeprecatedPath()
                )
            );
    }
```

Reply:
Unfortunately the `RestController` is missing quite a few features:
 - no way to confirm something was added
 - no way to remove something (other than by overwriting with something else)

There's probably a future feature improvement to the controller itself that could help with some of these...

Reply:
May be open an issue and we can pick it up to improve OpenSearch `main`